app/models/minion.rb

```diff
@@ -4,18 +4,43 @@
 
 # Minion represents the minions that have been registered in this application.
 class Minion < ApplicationRecord
+  class NotEnoughMinions < StandardError; end
+  class CouldNotAssignRole < StandardError; end
+
   enum role: [:master, :minion]
 
   validates :hostname, uniqueness: true
 
+  # This method is used to assign the specified roles to Minions which do not
+  # have a roles already assigned (available minions).
+  # roles param are the roles we absolutely have to assign. If we can't assign
+  # one of those, the method will raise.
+  # default_role param can be set if we want all the rest of the available
+  # minions to get a default role.
+  def self.assign_roles(roles: [], default_role: nil)
+    minions = Minion.where(role: nil)
+
+    # only the needed number of minions or all if we have a default role
+    minions = minions.limit(roles.size) unless default_role
+
+    raise NotEnoughMinions if minions.count < roles.size
+
+    minions.find_each do |minion|
+      unless minion.assign_role(roles.pop || default_role)
+        raise CouldNotAssignRole
```

Comment:
If one minion fails to assign the role, we stop the loop because we raise the exception. Maybe we could do something different here, maybe a more friendly failure, where we try to do our best to assign roles even if some minion fails to be assigned the role.